swarm/runtime/statsdb/ingestion.py

```
@@ -552,7 +552,17 @@ def ingest_events(self, events: List[Dict[str, Any]], run_id: str) -> int:
         # Set ingestion context to allow record_* calls
         _ingestion_context.active = True
         try:
-            return self._ingest_events_internal(events, run_id)
+            # Optimize: use a single transaction for the entire batch
+            # This significantly reduces I/O overhead compared to per-row commits
+            with self._lock:
+                self.connection.execute("BEGIN TRANSACTION")
+                try:
+                    result = self._ingest_events_internal(events, run_id)
+                    self.connection.execute("COMMIT")
+                    return result
```

Ruff (0.15.0)

[warning] Consider moving this statement to an `else` block

(TRY300)